### ISSUE TITLE
Can't run .pb file on model_analyzer

### ISSUE THREAD

**Issue by happyseone** [triton-inference-server/model_analyzer#111]
Hello, I'm new at using model_analyzer.

I got a result of model_analyzer by doing QuickStart example, but I was not able to get the result on .pb file model especially.
I also check that I can execute perf-analyzer with these models. However, it is not possible to run in model-analyzer with the same setting.

Additionally, I have a question about the results of model analyzer.
1. First, I run these codes to see the results of QuickStart.
(1) Run the Triton Inference Server first
`docker run --gpus=1 --rm -p8000:8000 -p8001:8001 -p8002:8002 -v /home/model_analyzer/examples/quick-start:/home/model_analyzer/examples/quick-start nvcr.io/nvidia/tritonserver:20.11-py3 tritonserver --model-control-mode=explicit --model-repository=/home/model_analyzer/examples/quick-start/`
(2) Open the model-analyzer container on the docker
`docker run -it --privileged --rm --gpus all \
        -v /var/run/docker.sock:/var/run/docker.sock \
        -v /home/model_analyzer/examples/quick-start:/home/model_analyzer/examples/quick-start \
        --net=host --name model-analyzer \
        model-analyzer /bin/bash`
(3)
`model-analyzer -m /home/model_analyzer/examples/quick-start -n add_sub --triton-launch-mode=remote --export-path=analysis_results`

(4) Results
Server Only:
Model           GPU ID   GPU Memory Usage (MB)   GPU Utilization (%)   GPU Power Usage (W)
triton-server   0        626.0                   0.0                   15.0

Models (Inference):
Model     Batch   Concurrency   Model Config Path   Instance Group   Dynamic Batcher Sizes   Satisfies Constraints   Throughput (infer/sec)   p99 Latency (ms)   RAM Usage (MB)
add_sub   1       16            add_sub             1/GPU            Disabled                Yes                     12387.4                  1.5                0.0
add_sub   1       4             add_sub             1/GPU            Disabled                Yes                     12164.8                  0.4                0.0
add_sub   1       8             add_sub             1/GPU            Disabled                Yes                     12083.8                  0.7                0.0
add_sub   1       2             add_sub             1/GPU            Disabled                Yes                     11931.6                  0.2                0.0
add_sub   1       1             add_sub             1/GPU            Disabled                Yes                     3057.6                   1.4                0.0

Models (GPU Metrics):
Model     GPU ID   Batch   Concurrency   Model Config Path   Instance Group   Dynamic Batcher Sizes   Satisfies Constraints   GPU Memory Usage (MB)   GPU Utilization (%)   GPU Power Usage (W)
add_sub   0        1       16            add_sub             1/GPU            Disabled                Yes                     624.0                   8.2                   31.8
add_sub   0        1       4             add_sub             1/GPU            Disabled                Yes                     624.0                   8.1                   31.5
add_sub   0        1       8             add_sub             1/GPU            Disabled                Yes                     624.0                   8.1                   31.7
add_sub   0        1       2             add_sub             1/GPU            Disabled                Yes                     624.0                   8.1                   31.6
add_sub   0        1       1             add_sub             1/GPU            Disabled                Yes                     624.0                   2.4                   30.9

What I want to know is that how GPU memory usage and GPU Utilization are almost same from Concurrency 2 to Concurrency 16. I think that concurrency is the execution of many models in parallel. Could you give the more explanation of concurrency please?

2. I want to run other models on model-analyzer. In the each model folder,  it has '1' folder where model.pb file is in and config.pbtxt and output_labels.txt. These model folders are in the same directory of QuickStart.
('add_sub', 'apple', 'new' location : /home/model_analyzer/examples/quick-start)
My custom model is in the 'apple' folder and [simple_identity](https://github.com/triton-inference-server/server/tree/master/docs/examples/model_repository/simple_identity) model is in the 'new' folder.

Config file(config.pbtxt) explanation is below:

`name: "simple_identity"
platform: "tensorflow_savedmodel"
max_batch_size: 8

input [
  {
    name: "INPUT0"
    data_type: TYPE_STRING
    dims: [ -1 ]
  }
]
output [
  {
    name: "OUTPUT0"
    data_type: TYPE_STRING
    dims: [ -1 ]
    label_filename: "output0_labels.txt"
  }
]`

The error message what I've got was 
`root@d1:/# model-analyzer -m /home/model_analyzer/examples/quick-start -n new --triton-launch-mode=remote --export-path=analysis_results
2021-04-08 07:49:53.278 INFO[entrypoint.py:288] Triton Model Analyzer started: config={'model_repository': '/home/model_analyzer/examples/quick-start', 'model_names': [{'model_name': 'new', 'objectives': {'perf_throughput': 10}, 'parameters': {'batch_sizes': [1], 'concurrency': []}}], 'objectives': {'perf_throughput': 10}, 'constraints': {}, 'batch_sizes': [1], 'concurrency': [], 'perf_analyzer_timeout': 600, 'perf_analyzer_cpu_util': 80.0, 'run_config_search_max_concurrency': 1024, 'run_config_search_max_instance_count': 5, 'run_config_search_disable': False, 'run_config_search_max_preferred_batch_size': 16, 'export': True, 'export_path': 'analysis_results', 'summarize': True, 'filename_model_inference': 'metrics-model-inference.csv', 'filename_model_gpu': 'metrics-model-gpu.csv', 'filename_server_only': 'metrics-server-only.csv', 'max_retries': 100, 'duration_seconds': 5, 'monitoring_interval': 0.01, 'client_protocol': 'grpc', 'perf_analyzer_path': 'perf_analyzer', 'perf_measurement_window': 5000, 'perf_output': False, 'triton_launch_mode': 'remote', 'triton_docker_image': 'nvcr.io/nvidia/tritonserver:21.02-py3', 'triton_http_endpoint': 'localhost:8000', 'triton_grpc_endpoint': 'localhost:8001', 'triton_metrics_url': 'http://localhost:8002/metrics', 'triton_server_path': 'tritonserver', 'triton_output_path': None, 'triton_server_flags': {}, 'log_level': 'INFO', 'gpus': ['all'], 'output_model_repository_path': './output_model_repository', 'override_output_model_repository': False, 'config_file': None, 'inference_output_fields': ['model_name', 'batch_size', 'concurrency', 'model_config_path', 'instance_group', 'dynamic_batch_sizes', 'satisfies_constraints', 'perf_throughput', 'perf_latency', 'cpu_used_ram'], 'gpu_output_fields': ['model_name', 'gpu_id', 'batch_size', 'concurrency', 'model_config_path', 'instance_group', 'dynamic_batch_sizes', 'satisfies_constraints', 'gpu_used_memory', 'gpu_utilization', 'gpu_power_usage'], 'server_output_fields': ['model_name', 'gpu_id', 'gpu_used_memory', 'gpu_utilization', 'gpu_power_usage'], 'plots': [{'name': 'throughput_v_latency', 'title': 'Throughput vs. Latency', 'x_axis': 'perf_latency', 'y_axis': 'perf_throughput', 'monotonic': True}, {'name': 'gpu_mem_v_latency', 'title': 'GPU Memory vs. Latency', 'x_axis': 'perf_latency', 'y_axis': 'gpu_used_memory', 'monotonic': False}], 'top_n_configs': 3}
2021-04-08 07:49:53.280 INFO[entrypoint.py:79] Using remote Triton Server...
2021-04-08 07:49:53.280 WARNING[entrypoint.py:82] GPU memory metrics reported in the remote mode are not accuracte. Model Analyzer uses Triton explicit model control to load/unload models. Some frameworks do not release the GPU memory even when the memory is not being used. Consider using the "local" or "docker" mode if you want to accurately monitor the GPU memory usage for different models.
2021-04-08 07:49:53.280 WARNING[entrypoint.py:89] Config sweep parameters are ignored in the "remote" mode because Model Analyzer does not have access to the model repository of the remote Triton Server.
2021-04-08 07:49:53.337 INFO[driver.py:236] init
2021-04-08 07:49:54.404 INFO[entrypoint.py:327] Starting perf_analyzer...
2021-04-08 07:49:54.404 INFO[analyzer.py:82] Profiling server only metrics...
2021-04-08 07:49:55.431 INFO[gpu_monitor.py:73] Using GPU(s) with UUID(s) = { GPU-c8fdb676-2c11-669a-4cff-f300b28eb26a } for the analysis.
2021-04-08 07:49:56.464 INFO[run_search.py:155] Will sweep only through the concurrency values...
2021-04-08 07:49:56.464 INFO[run_search.py:262] Concurrency set to 1.
2021-04-08 07:49:56.468 INFO[client.py:82] Model new load failed: [StatusCode.INTERNAL] failed to load 'new', no version is available
2021-04-08 07:50:01.584 INFO[client.py:143] Model readiness failed for model new. Error None
Traceback (most recent call last):
  File "/usr/local/bin/model-analyzer", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/entrypoint.py", line 328, in main
    analyzer.run()
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/analyzer.py", line 95, in run
    run_config_generator = RunConfigGenerator(
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/config/run/run_config_generator.py", line 65, in __init__
    self._generate_run_configs()
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/config/run/run_config_generator.py", line 290, in _generate_run_configs
    self._generate_run_config_for_model_sweep(
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/config/run/run_config_generator.py", line 229, in _generate_run_config_for_model_sweep
    model_config = ModelConfig.create_from_triton_api(
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/triton/model/model_config.py", line 111, in create_from_triton_api
    model_config_dict = client.get_model_config(model_name, num_retries)
  File "/usr/local/lib/python3.8/dist-packages/model_analyzer/triton/client/grpc_client.py", line 54, in get_model_config
    model_config_dict = self._client.get_model_config(model_name,
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/grpc/__init__.py", line 476, in get_model_config
    raise_error_grpc(rpc_error)
  File "/usr/local/lib/python3.8/dist-packages/tritonclient/grpc/__init__.py", line 61, in raise_error_grpc
    raise get_error_grpc(rpc_error) from None
tritonclient.utils.InferenceServerException: [StatusCode.UNAVAILABLE] Request for unknown model: 'new' is not found`

I can run perf-analyzer with the same setting, but can't run in model-analyzer.

**Comment by Tabrizian** [comment]
@happyseone Concurrency is a parameter designed to adjust the request load on Triton Server. You can read more about it below:

https://github.com/triton-inference-server/server/blob/master/docs/perf_analyzer.md#request-concurrency

Regarding the error that you are seeing, it looks like `new` model is not in the model repository folder. Can you share the output of `tree /home/model_analyzer/examples/quick-start`? It should look like below:

```
/home/model_analyzer/examples/quick-start
├── add_sub
├── new
└── apple
```

The `new` model should be in the same model repository as other models.

> What I want to know is that how GPU memory usage and GPU Utilization are almost same from Concurrency 2 to Concurrency 16. I think that concurrency is the execution of many models in parallel.

If you don't have dynamic batching enabled, GPU memory usage will likely not change throughout different concurrency values. The reason is there is no need for new memory allocations. But if you have dynamic batching enabled GPU memory usage will change.

Regarding the GPU utilization, if you have a single instance of your model all requests to that model will be serialized and because of this, the GPU utilization will be the same.

I also noticed that you are using `remote` mode. Did you know about the `docker` and `local` mode? In those modes, model analyzer will change some parameters of your model that may lead to better GPU utilization. It may take longer, but is better suited if you want to increase your model throughput/decrease latency.

**Comment by happyseone** [comment]
Thank you very much. The problem was resolved. The cause of the problem was that model_name in the config file and folder name were different.

By the way, I have another issue to run model-analyzer.
My input shape in config file is [-1, 2] because my every input can have different length.
In the config file, Instance group count is 4 and dynamic batching size is 4.
When I run the model-analyzer with the input shape [-1,2], the error message is below.[My model name is SE-DKT+]

`2021-04-09 00:39:30.928 INFO[perf_analyzer.py:135] Running perf_analyzer ['perf_analyzer', '-m', 'SE-DKT+', '-b', '1', '-u', 'localhost:8001', '-i', 'grpc', '--measurement-interval', '5000', '--concurrency-range', '1'] failed with exit status 1 : error: failed to create concurrency manager: input inputs contains dynamic shape, provide shapes to send along with the request

2021-04-09 00:39:31.900 WARNING[result_manager.py:278] Requested top 3 configs, but none satisfied constraints. Showing available constraint failing configs for this model.
2021-04-09 00:39:31.900 WARNING[result_manager.py:283] Requested top 3 failing configs, but found only 0. Showing all available constraint failing configs for this model.
2021-04-09 00:39:31.900 WARNING[legend.py:1225] No handles with labels found to put in legend.
2021-04-09 00:39:31.901 WARNING[legend.py:1225] No handles with labels found to put in legend.`

Then, I can run model-analyzer after I change many types of input shape such as [10,2], [20,2], [200, 2] in config file.
I'm curious that how model-analyzer can calculate GPU-memory/Util and throughput with different input shape.
For example, when I change input shape as [20, 2] in config file but some input shapes are bigger than [20,2], because I run the model with dynamic shape inputs.
Does model-analyzer only use the part of input data [20,2] and waste the other parts?

Also, Let's say I change input shape as [200,2] in config file but some input shapes are smaller than [200,2] such as [20,2]
Does model-analyzer make the zero padding in these kinds of input data to calculate the GPU things?

**Comment by aramesh7** [comment]
@happyseone With regards to your error, that seems to be an error from perf_analyzer. perf_analyzer has a specific command line option to specify input shapes dynamically. You can read about providing input shapes to perf_analyzer [here](https://github.com/triton-inference-server/server/blob/master/docs/perf_analyzer.md#input-data). 

Model Analyzer allows you to pass the above flags (such as shape) to the perf_analyzer instances launched as well via the config. See [these docs](https://github.com/triton-inference-server/model_analyzer/blob/main/docs/config.md#perf-analyzer-flags) for more info.

Neither model analyzer, nor perf_analyzer nor Triton make any changes to your data. If you fix the input shape in the config, thats what the server will expect. If the shape is dynamic, the server will expect that the request to contain information about the input shape via the `--shape` argument to perf_analyzer.

**Comment by happyseone** [comment]
Thank you for the explanation. 
I want to change or make the yaml file for new configuration, but I can't find where the yaml file is.
Unfortunately, Changing <perf-analyzer-flags> are only available in the config.yaml file I guess. 

`root@d1:/home/model_analyzer# find . -name "*.yaml"
./qa/L0_custom_flags/config.yaml
./helm-chart/values.yaml
./helm-chart/templates/job.yaml
./helm-chart/Chart.yaml`

Could you let me know where I make or put the yaml file?
My workspace is just cloned by this github except my model files, so I can easily follow your instruction.

**Comment by Tabrizian** [comment]
If you want to use the YAML file, you need to create it by yourself. You can create it wherever you want and run model analyzer using `-f` flag:

`model-analyzer -f <path-to-config.yml>`

Documentation is available in the link below:

https://github.com/triton-inference-server/model_analyzer/blob/main/docs/config.md#example-1

This example is probably more than what you need but shows lots of different config options available in model analyzer.

**Comment by happyseone** [comment]
I can run the model analyzer with my config.yaml file but I'm very sorry to write another issue.
I got the same error msg on the perf analyzer and model analyzer with dynamic input shape when the first dimension is '-1' 

To run the perf analyzer,
`perf_client -m SE-DKT+ --percentile=95 --concurrency-range 1:8 --shape inputs:-1,2`

To run the model analyzer, I write the config.yml file like this.
model_repository: /home/model_analyzer/examples/quick-start
run_config_search_disable: false
triton_launch_mode: remote
export_path: /home/model_analyzer/examples/quick-start/metrics

model_names:
  SE-DKT+:
    parameters:
        batch_sizes: 1, 2, 4, 8, 16, 32
    perf_analyzer_flags:
        percentile: 95
        shape: inputs:-1,2
    model_config_parameters:
        dynamic_batching:
            preferred_batch_size: [[4]]
        instance_group:
        -
            -
                kind: KIND_CPU
                count: 4

My model config.pbtxt file is below
name: "SE-DKT+"
platform: "tensorflow_savedmodel"
max_batch_size: 32

input [
  {
    name: "inputs"
    data_type: TYPE_INT32
    dims: [-1 , 2]  #model-analyzer can be run when the first dimension is not '-1' 
  },
  {
    name: "input_c_state"
    data_type: TYPE_FP32
    dims: [ 200]
  },
  {
    name: "input_h_state"
    data_type: TYPE_FP32
    dims: [ 200]
  }
output [
  {
    name: "preds"
    data_type: TYPE_FP32
    dims: [ 544 ]

  },
  {
    name: "output_c_state"
    data_type: TYPE_FP32
    dims: [ 200 ]
  },
  {
    name: "output_h_state"
    data_type: TYPE_FP32
    dims: [ 200 ]
  }
]
instance_group [
  {
    count: 4
    kind: KIND_GPU
  }
]
dynamic_batching {
preferred_batch_size : [4]


the error message is

root@d1:/home/model_analyzer/examples/quick-start# model-analyzer -f /home/model_analyzer/examples/quick-start/config.yaml
2021-04-13 03:39:18.864 ERROR[entrypoint.py:285] Model Analyzer encountered an error: Export path /home/model_analyzer/examples/quick-start/metrics is not a directory.
root@d1:/home/model_analyzer/examples/quick-start# mkdir metrics
root@d1:/home/model_analyzer/examples/quick-start# model-analyzer -f /home/model_analyzer/examples/quick-start/config.yaml
2021-04-13 03:39:28.412 INFO[entrypoint.py:288] Triton Model Analyzer started: config={'model_repository': '/home/model_analyzer/examples/quick-start', 'model_names': [{'model_name': 'SE-DKT+', 'objectives': {'perf_throughput': 10}, 'parameters': {'batch_sizes': [1, 2, 4, 8, 16, 32], 'concurrency': []}, 'model_config_parameters': {'dynamic_batching': [{'preferred_batch_size': [[4]]}], 'instance_group': [[{'kind': ['KIND_CPU'], 'count': [4]}]]}, 'perf_analyzer_flags': {'percentile': '95', 'shape': 'inputs:-1,2'}}], 'objectives': {'perf_throughput': 10}, 'constraints': {}, 'batch_sizes': [1], 'concurrency': [], 'perf_analyzer_timeout': 600, 'perf_analyzer_cpu_util': 80.0, 'run_config_search_max_concurrency': 1024, 'run_config_search_max_instance_count': 5, 'run_config_search_disable': False, 'run_config_search_max_preferred_batch_size': 16, 'export': True, 'export_path': '/home/model_analyzer/examples/quick-start/metrics', 'summarize': True, 'filename_model_inference': 'metrics-model-inference.csv', 'filename_model_gpu': 'metrics-model-gpu.csv', 'filename_server_only': 'metrics-server-only.csv', 'max_retries': 100, 'duration_seconds': 5, 'monitoring_interval': 0.01, 'client_protocol': 'grpc', 'perf_analyzer_path': 'perf_analyzer', 'perf_measurement_window': 5000, 'perf_output': False, 'triton_launch_mode': 'remote', 'triton_docker_image': 'nvcr.io/nvidia/tritonserver:21.02-py3', 'triton_http_endpoint': 'localhost:8000', 'triton_grpc_endpoint': 'localhost:8001', 'triton_metrics_url': 'http://localhost:8002/metrics', 'triton_server_path': 'tritonserver', 'triton_output_path': None, 'triton_server_flags': {}, 'log_level': 'INFO', 'gpus': ['all'], 'output_model_repository_path': './output_model_repository', 'override_output_model_repository': False, 'config_file': '/home/model_analyzer/examples/quick-start/config.yaml', 'inference_output_fields': ['model_name', 'batch_size', 'concurrency', 'model_config_path', 'instance_group', 'dynamic_batch_sizes', 'satisfies_constraints', 'perf_throughput', 'perf_latency', 'cpu_used_ram'], 'gpu_output_fields': ['model_name', 'gpu_id', 'batch_size', 'concurrency', 'model_config_path', 'instance_group', 'dynamic_batch_sizes', 'satisfies_constraints', 'gpu_used_memory', 'gpu_utilization', 'gpu_power_usage'], 'server_output_fields': ['model_name', 'gpu_id', 'gpu_used_memory', 'gpu_utilization', 'gpu_power_usage'], 'plots': [{'name': 'throughput_v_latency', 'title': 'Throughput vs. Latency', 'x_axis': 'perf_latency', 'y_axis': 'perf_throughput', 'monotonic': True}, {'name': 'gpu_mem_v_latency', 'title': 'GPU Memory vs. Latency', 'x_axis': 'perf_latency', 'y_axis': 'gpu_used_memory', 'monotonic': False}], 'top_n_configs': 3}
2021-04-13 03:39:28.414 INFO[entrypoint.py:79] Using remote Triton Server...
2021-04-13 03:39:28.414 WARNING[entrypoint.py:82] GPU memory metrics reported in the remote mode are not accuracte. Model Analyzer uses Triton explicit model control to load/unload models. Some frameworks do not release the GPU memory even when the memory is not being used. Consider using the "local" or "docker" mode if you want to accurately monitor the GPU memory usage for different models.
2021-04-13 03:39:28.414 WARNING[entrypoint.py:89] Config sweep parameters are ignored in the "remote" mode because Model Analyzer does not have access to the model repository of the remote Triton Server.
2021-04-13 03:39:28.468 INFO[driver.py:236] init
2021-04-13 03:39:29.600 INFO[entrypoint.py:327] Starting perf_analyzer...
2021-04-13 03:39:29.601 INFO[analyzer.py:82] Profiling server only metrics...
2021-04-13 03:39:30.627 INFO[gpu_monitor.py:73] Using GPU(s) with UUID(s) = { GPU-c8fdb676-2c11-669a-4cff-f300b28eb26a } for the analysis.
2021-04-13 03:39:31.661 INFO[run_search.py:155] Will sweep only through the concurrency values...
2021-04-13 03:39:31.661 INFO[run_search.py:262] Concurrency set to 1.
2021-04-13 03:39:31.694 INFO[client.py:80] Model SE-DKT+ loaded.
2021-04-13 03:39:31.697 INFO[client.py:104] Model SE-DKT+ unloaded.
2021-04-13 03:39:31.776 INFO[client.py:80] Model SE-DKT+ loaded.
2021-04-13 03:39:31.777 INFO[run_config_generator.py:177] Profiling model SE-DKT+...
2021-04-13 03:39:32.808 INFO[gpu_monitor.py:73] Using GPU(s) with UUID(s) = { GPU-c8fdb676-2c11-669a-4cff-f300b28eb26a } for the analysis.
2021-04-13 03:39:37.861 INFO[perf_analyzer.py:135] Running perf_analyzer ['perf_analyzer', '-m', 'SE-DKT+', '-b', '32', '-u', 'localhost:8001', '-i', 'grpc', '--measurement-interval', '5000', '--concurrency-range', '1', '--percentile', '95', '--shape', 'inputs:-1,2'] failed with exit status 1 : error: input shape must be > 0

I put various input shapes but when the first dimension of input shape is '-1' for dynamic shape, I got the error message "error: input shape must be > 0"

I feel very sorry to ask several times, but I want to the problem to be solved. 
Thank you for your instructions.

**Comment by Tabrizian** [comment]
@happyseone It looks like the problem is with the `shape` field. When you are providing the shape field, you need to avoid specifying `-1` as the dimension.
